pycparser/c_ast.py

```diff
@@ -510,6 +510,22 @@ def __iter__(self):
 
     attr_names = ('name', )
 
+class Atomic(Node):
```

Comment:
There are two issues here:

1. You should not modify `c_ast.py` directly - it's an auto-generated file. Modify `_c_ast.cfg` instead and rerun `_ast_gen.py`. The README mentions this briefly
2. It's not clear to me why Atomic is a Node at all, if it's only a type qualifier or specifier (like `const` or `static`). Can't it be handled the same way?

Reply:
1. Will check.
2. It is both a specifier and a qualifier, this node is to handle a specifier.

Reply:
Forgive my ignorance, but how can I run `c_ast.py`?

Reply:
Specifiers are handled without nodes. See how `register` is handled, for example. BTW, `_Thread_local` is also a specifier (per https://en.cppreference.com/w/c/language/storage_duration) so you would have to follow the same path there.

Reply:
Sorry, but how can `_Atomic(_Atomic(int) *)` be reconstructed without a node?

Reply:
> Forgive my ignorance, but how can I run `_ast_gen.py`?

My bad, the instructions are not clear about this. I've updated the `README` to be more clear. Please LMK if it works for you.

Reply:
Thank you, the instructions did indeed work for me, and I made a commit with the changes. As for `_Atomic`, I do not believe there is a way to handle them without a Node. Let me know if you think I am wrong.

Reply:
I'm not deeply familiar with C11 or `_Atomic`, but it's not clear to me how this is different from `const int * const p`, where both the integer and the pointer are const. Isn't your example similar (both the integer and the pointer are atomic)?

Is the issue here because `_Atomic` could be written both ways? From reading the docs it seems they both compile to the same thing, and there's no semantic difference (except that as a specifier there are more restrictions - but pycparser typically assumes it's compiling correct code anyway).

Reply:
`_Atomic(x)` and `_Atomic x` are different things and the former is not equivalent to the latter. Several C11 specifiers, like `_Alignas(x)` or `_Atomic(x)` have arguments, and thus require nodes.

Reply:
Reading http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1485.pdf, it says:

> 
> "The _Atomic keyword also can also be used in the form _Atomic(T), where T is a type, as a type specifier equiva-
> lent to _Atomic T."
> 

So semantically, they seem equivalent. The right way to have specifiers in pycparser is as the `quals` attribute on declarations, and having `Atomic` as a node will add a second way to represent something equivalent.

One issue I can see is that having it as a node makes *parsing* nested qualifiers like the one you've shown easier. This is something to think about. What I mean is that, if you have:

`_Atomic(<some decl>)`, you can return the node for `<some decl>` with the atomic qualifier added, and so on recursively. Would that work, without a special node?

FWIW I think `_Alignas` is different, because its argument has semantic meaning beyond wrapping another type.

Reply:
>> "The _Atomic keyword also can also be used in the form _Atomic(T), where T is a type, as a type specifier equivalent to _Atomic T."
>
> So semantically, they seem equivalent.

I think there are two sides here:

- It works only in one direction. I.e. every pair of qualifier-type `_Atomic T` can be written as specifier-type `_Atomic(T)`. It is not possible to do it vice-versa, i.e. not every pair of specifier-type  `_Atomic(T)` can be written as qualifier-type `_Atomic T`.
- For the AST I believe it may be important to distinguish the pair type. For instance, one can make style suggestions in this case like "Rewrite it this way or that way".

> `_Atomic(<some decl>)`, you can return the node for `<some decl>` with the atomic qualifier added, and so on recursively. Would that work, without a special node?

I do not believe it can work, because of the following example:

```C
_Atomic(int *) a;
_Atomic(int) *b;
_Atomic int *c;
```

`a` is an atomic storing `int` pointers, while `b` and `c` are pointers to atomic `int`. In my opinion, for this very reason I cannot put `_Atomic` to something like `quals` attribute in case it is a specifier, as it clearly defines a type hierarchy. I believe the closest what you have in the code is the `enum` specifier.

> FWIW I think `_Alignas` is different, because its argument has semantic meaning beyond wrapping another type.

I am not sure. Currently I am implementing `_Alignas` and have difficulties understanding how the grammar is applied. I pushed a change into https://github.com/vit9696/pycparser/tree/alignas, might you give me a hint what I am missing? It suggests that I have two types (`Invalid multiple types specified`), but I do not understand why this happens.